Support for SLEPc eigensolver [slepc-dev]

jetremblay opened the pull request:
Hi,

I implement support for the SLEPc eigensolver library. This is heavily inspired from the current PETSc code, and uses internally PetscParMatrix. Some very basic options are supported, and other options can be specified in an options file (similar to PETSc).

I ported ex11p, and added ex28p which solves the 2D Maxwell eigenproblem in a dielectric waveguide, surrounded by metallic walls.

This is currently WIP:
- [x] Test building with regular Make system
- [x] Improve ex28p solver options and documentation
- [x] ~Add SLEPc to CI build~ CI doesn't test for PETSc so I didn't add SLEPc
<!--GHEX{"id":1487,"author":"jetremblay","editor":"tzanio","reviewers":["stefanozampini","mlstowell","jandrej"],"assignment":"2020-05-20T08:23:16-07:00","approval":"2020-06-20T19:02:08.428Z","merge":"2020-07-05T01:54:38.503Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1487](https://github.com/mfem/mfem/pull/1487) | @jetremblay | @tzanio | @stefanozampini + @mlstowell + @jandrej | 05/20/20 | 06/20/20 | 07/04/20 | |
<!--ELBATXEHG-->

tzanio commented:
Can you please

1. Resolve the Travis errors (the code should build fine without `MFEM_USE_SLEPC`)
2. Mark the PR with the `ready-for-review` when ready, see [CONTRIBUTING.md](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#quick-summary)

Cheers,
Tzanio

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,342 @@
+   const Operator       *oA = dynamic_cast<const Operator *>(&op);
+   bool delete_pA = false;
+
+   if (!pA)
```
Since SLEPc is based on PETSc, I don't see the reason why we need to convert here.
It is responsibility of the users to pass a `PetscParMatrix`. There are plenty of ways to construct it, and it should be done by users, not by MFEM. I suggest you change the interface with `const PetscParMatrix&` as argument. 
This is different for the linear and nonlinear solvers, since you want to have a similar API for MFEM and PETSc based solvers. @v-dobrev and @tzanio ?

jetremblay replied:
I can do that. In comparison, the Hypre LOBPCG takes `Operator&` as argument and casts to a `HypreParMatrix`, but I am not really trying to match exactly the LOBPCG interface.

jandrej replied:
I agree with @stefanozampini. The PETSc interface gives the user more flexibility to decide whether he wants to pass a nested or monolithic matrix etc.

jetremblay replied:
I changed the interface to only take `PetscParMatrix`, and changed ex11p accordingly.

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,342 @@
+void SlepcEigenSolver::GetEigenvalue(unsigned int i, double & lr,
+                                     double & lc) const
+{
+   MFEM_VERIFY(i < _num_conv,"Eigenvalue not computed");
```
I would let SLEPc fail here, not MFEM

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,342 @@
+                                   bool wrap)
+{
+   clcustom = false;
+   _tol = PETSC_DEFAULT;
```
I do not see the reason why you want to store `_tol`, `_max_its` and `_num_conv` in the class since you are not using them.

jetremblay replied:
It seems like `EPSSetTolerances` is different from e.g. `KSPSetTolerances`, in that if you pass PETSC_DEFAULT it will reset that parameter to default, instead of ignoring it. Because of this we need to keep an internal state if we want to change only one at a time. Do you prefer that I simply wrap `EPSSetTolerances`?

stefanozampini replied:
In principle, the MFEM API should only be a proxy for EPS API.
You can call EPSGetTolerances inside MFEM API to preserve any already set value. 
If we want to keep SetTol and SetMaxIt separate, what should we do? @joseeroman?

joseeroman replied:
If you want to change only one of the arguments, call `EPSGetTolerances` first and then `EPSSetTolerances` with one value modified.

stefanozampini replied:
And this will work even if the parameter has not been set before? Keeping the state of the EPS solver intact and not having any unexpected side-effect? Should we call EPSSetTolerance(eps,PETSC_DEFAULT,PETSC_DEFAULT) at creation time to avoid this?

joseeroman replied:
Now that I look at the code, the two parameters have different behaviour, the tolerance is set initially to `PETSC_DEFAULT` but the maximum iterations have the value 0 to indicate uninitialized. So `EPSGetTolerances` followed by `EPSSetTolerances` will work for the former but not the latter. Do you want me to fix it?

stefanozampini replied:
You mean you need to fix it in SLEPc? If zero is a valid value for maxit, then what's wrong about calling `EPSSetTolerance(eps,PETSC_DEFAULT,PETSC_DEFAULT)` at object creation and then use the get-first-then-set approach for the tolerances?

joseeroman replied:
Yes, I mean fixing in SLEPc.
No, maxit=zero is not allowed in `EPSSetTolerances`, that is the problem.

stefanozampini replied:
Oh, I see now. If 0 will always indicate uninitialized maxits in SLEPc, we can do

```
EPSGetTolerances(eps,&t,&m)
if (m==0) m = PETSC_DECIDE;
EPSSetTolerances(eps,t,m)
```
We don't need to by tight to a specific version of SLEPc having this fix.

joseeroman replied:
Correct. Anyway, I will fix it in slepc-maint

jetremblay replied:
Thanks for the clarifications, I will add that so it will work around the previous versions of SLEPc.

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,342 @@
+      delete VC;
+      VR = VC = NULL;
+   }
+   operatorset = true;
```
`operatorset` is used in the `PetscLinearSolver` class because we can in principle reuse the `KSP` with different matrix sizes, but we need to call `KSPReset` on it. I'm not sure we can do the same with the `EPS` object. I suggest you remove this logic.

jetremblay replied:
Another difference between `EPS` and `KSP`, is that `EPSSetOperators` calls `EPSReset` if necessary. However we still need to reset VC and VR is we reuse the `EPS`. I could remove the logic and still always delete `VC` and `VR`?

stefanozampini replied:
Yes, delete and recreate the vectors placeholders at each call to SetOperators then.

jetremblay replied:
Done

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,342 @@
+
+   ierr = EPSSolve(eps); PCHKERRQ(eps,ierr);
+
+   ierr = EPSGetConverged(eps,&_num_conv); PCHKERRQ(eps,ierr);
```
The less API we use, the more general the interface. Remove this call

stefanozampini commented on linalg/slepc.hpp:
```diff
@@ -0,0 +1,127 @@
+#include "../config/config.hpp"
+
+#ifdef MFEM_USE_SLEPC
+#ifdef MFEM_USE_PETSC
```
MFEM_USE_SLEPC should imply MFEM_USE_PETSC in the building process. No need to check it here

stefanozampini commented on config/config.mk.in:
```diff
@@ -37,6 +37,7 @@ MFEM_USE_GINKGO        = @MFEM_USE_GINKGO@
 MFEM_USE_GNUTLS        = @MFEM_USE_GNUTLS@
 MFEM_USE_NETCDF        = @MFEM_USE_NETCDF@
 MFEM_USE_PETSC         = @MFEM_USE_PETSC@
+MFEM_USE_SLEPC          = @MFEM_USE_SLEPC@
```
align

stefanozampini commented on config/defaults.mk:
```diff
@@ -125,6 +125,7 @@ MFEM_USE_GINKGO        = NO
 MFEM_USE_GNUTLS        = NO
 MFEM_USE_NETCDF        = NO
 MFEM_USE_PETSC         = NO
+MFEM_USE_SLEPC          = NO
```
align

stefanozampini commented on config/defaults.mk:
```diff
@@ -275,6 +276,19 @@ ifeq ($(PETSC_FOUND),YES)
+ifeq ($(SLEPC_FOUND),YES)
+    SLEPC_OPT := $(shell set -n "s/$(SLEPC_INC_VAR) = *//p" $(SLEPC_VARS))
+    SLEPC_LIB := $(shell set -n "s/$(SLEPC_INC_LIB) = *//p" $(SLEPC_VARS))
+    PETSC_LIB := -Wl,-rpath,$(abspath $(SLEPC_DIR))/lib\
```
Are you sure about it?

jetremblay replied:
Probably a copy paste error. I will check this.

jetremblay replied:
I fixed it, and tested with make.

jandrej commented on config/cmake/modules/FindSLEPc.cmake:
```diff
@@ -0,0 +1,278 @@
+# Assumes that PETSC_DIR and PETSC_ARCH has been set by
+# alredy calling find_package(PETSc)
+
+#=============================================================================
+# Copyright (C) 2010-2012 Garth N. Wells, Anders Logg and Johannes Ring
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in
+#    the documentation and/or other materials provided with the
+#    distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#=============================================================================
```
This has to be kept in mind. Might be necessary or not wanted for certain customer situations. Ping @acfisher @tzanio.

jetremblay replied:
I am not very familiar with cmake so I reused an existing file from another project. The FindPETSc.cmake has a similar header (this might be a separate issue). If you prefer I could write something based on https://gitlab.com/slepc/slepc/-/issues/19, but it would depend on pkg-config.

jetremblay replied:
What should I do with this?

v-dobrev replied:
This whole file can probably be simplified a lot by using something like this: https://github.com/mfem/mfem/blob/cc3a4475d298df742de0a98d9c4586c3fd02ef82/config/cmake/modules/FindSTRUMPACK.cmake#L12-L36

jandrej commented on examples/petsc/ex11p.cpp:
```diff
@@ -0,0 +1,439 @@
+//               mpirun -np 4 ex11p -m ../../data/mobius-strip.mesh -n 8
+//               mpirun -np 4 ex11p -m ../../data/klein-bottle.mesh -n 10
+//
+// Description:  This example code demonstrates the use of MFEM to solve the
+//               eigenvalue problem -Delta u = lambda u with homogeneous
+//               Dirichlet boundary conditions.
+//
+//               We compute a number of the lowest eigenmodes by discretizing
+//               the Laplacian and Mass operators using a FE space of the
+//               specified order, or an isoparametric/isogeometric space if
+//               order < 1 (quadratic for quadratic curvilinear mesh, NURBS for
+//               NURBS mesh, etc.)
+//
+//               The example highlights the use of the LOBPCG eigenvalue solver
+//               together with the BoomerAMG preconditioner in HYPRE, as well as
+//               optionally the SuperLU or STRUMPACK parallel direct solvers.
+//               Reusing a single GLVis visualization window for multiple
+//               eigenfunctions is also illustrated.
+//
+//               We recommend viewing Example 1 before viewing this example.
```
Maybe mention SLEPc in the description.

jetremblay replied:
Done

jandrej commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,352 @@
+// Sample runs:
+//    mpirun -np 4 ex28p -m ../../data/inline-tri.mesh --slepcopts rc_ex28p
+//
+// Description:  This example code solves a simple 2D dielectric waveguide problem
+//               corresponding to the generalized eigenvalue equation
+//          curl 1/mu curl exy - beta^2/mu (grad ez - exy) = k^2 epsilon exy
+//          beta^2 div 1/mu (grad exy - et) = beta^2 k^2 epsilon ez
+//               with essential boundary condition (corresponding to metallic walls).
+//               We discretize with Nédélec edge elements (transverge field exy)
+//               and piecewise continuous polynomials (longitudinal field ez).
```
Formatting

jetremblay replied:
Done

jandrej commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,352 @@
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+//#include <slepceps.h>
```
Remove

jetremblay replied:
Removed

jandrej commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,352 @@
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.AddOption(&slepcrc_file, "-slepcopts", "--slepcopts",
+                  "SLepcOptions file to use.");
```
Keep capitalization consistent

jetremblay replied:
Fixed

jandrej commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,352 @@
+   ConstantCoefficient u_r_func(1.0);
+   Vector e_r(2);
+   double k0 = M_PI*2/1.0;
+   e_r(0) = -pow(k0*1.0,2);//-k0^2*e_r
```
Align and format comment

jetremblay replied:
Done

jandrej commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,345 @@
+
+#include "slepc.h"
+
+// Error handling
+// Prints SLEPc's stacktrace and then calls MFEM_ABORT
+// We cannot use PETSc's CHKERRQ since it returns a PetscErrorCode
+#define PCHKERRQ(obj,err) do {                                                   \
+     if ((err))                                                                  \
+     {                                                                           \
+        PetscError(PetscObjectComm((PetscObject)(obj)),__LINE__,_MFEM_FUNC_NAME, \
+                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);                      \
+        MFEM_ABORT("Error in SLEPc. See stacktrace above.");                     \
+     }                                                                           \
+  } while(0);
+#define CCHKERRQ(comm,err) do {                                \
+     if ((err))                                                \
+     {                                                         \
+        PetscError(comm,__LINE__,_MFEM_FUNC_NAME,              \
+                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);    \
+        MFEM_ABORT("Error in SLEPc. See stacktrace above.");   \
+     }                                                         \
+  } while(0);
```
Can we create a common header file for this? I'd prefer another file instead of code duplication.

jetremblay replied:
I moved this to `petscinternals.hpp`. Let me know if this is what you had in mind!

jandrej commented on linalg/slepc.hpp:
```diff
@@ -0,0 +1,122 @@
+
+   /// Set maximum number of iterations
+   void SetMaxIter(int max_iter);
+   void SetNumModes(int num_eigs);
```
Description comment missing

jetremblay replied:
Commend added

stefanozampini commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,360 @@
+   pBzt = pBtz->Transpose();
+
+   PetscParMatrix *LHSOp = NULL, *RHSOp = NULL;
+   // We construct the BlockOperator and we then convert it to a
```
Are you using a MATNEST here with SLEPc?. If so, you should mention it here that you are preserving the blocked nature of the operator

jetremblay replied:
Yes this creates a MATNEST so only some solvers are available (e.g. Jacobi-Davidson method with Jacobi preconditioner). I can add a variation of the problem where the matrix is converted to MATAIJ and we can use e.g. shift-and-invert with a direct solver. I will improve the comments.

stefanozampini replied:
It would be nice to show users the flexibility of the PETSc matrix formats. If you add this, please also add a comment that converting from NEST to AIJ in sequential is reasonably fast, while doing in parallel may be a bit slower (In PETSc we currently do no handle fast conversion when the blocks are MPIAIJ; this could be added in the future if needed)

jetremblay replied:
You are correct, I also observed that the conversion takes longer in parallel. However, for my specific problem, the shift-and-invert method with a direct solver seems faster than using the Jacobi-Davidson method. Do you think using `HypreParMatrixFromBlocks` and then converting to a `PetscParMatrix` would be faster (potentially using a Hypre preconditioner in PETSc)?

stefanozampini replied:
If it is specifically written for mfem block operator, the answer is yes. However, my personal opinion is that it is best to keep the example as simple as possible. If you want users experiments with shift-and-invert, just add the conversion operator from the PetscParMatrix class (with the caveat about performances)

jetremblay replied:
Yes I should keep the example simple. Please see the latest version: https://github.com/mfem/mfem/blob/1855ec2993892ba4e5835bb3977278a93bfae852/examples/petsc/ex28p.cpp#L261-L274

stefanozampini commented on linalg/petscinternals.hpp:
```diff
@@ -0,0 +1,35 @@
+#ifndef MFEM_PETSCINTERNALS
+#define MFEM_PETSCINTERNALS
+
+// Error handling
```
without the proper include statements from petsc and MFEM headers this file is useless. Try compiling

```
#include "petscinternals.hpp"

int main()
{
int err = 0;
MPI_Comm comm;
CCHKERRQ(comm,err);
PetscObject obj;
PCHKERRQ(obj,err);
return 0;
}
```

jetremblay replied:
Good catch, fixed.

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,254 @@
+void SlepcEigenSolver::GetEigenvector(unsigned int i, Vector & vr,
+                                      Vector & vc) const
+{
+   if (!VR || !VC)
```
Should not needed, since you create the vectors at SetOperators time. Just add
```
MFEM_VERIFY(VR,"Missing right vector")
MFEM_VERIFY(VC,"Missing left vector")
```

jetremblay replied:
Done

stefanozampini commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,254 @@
+
+void SlepcEigenSolver::GetEigenvector(unsigned int i, Vector & vr) const
+{
+   if (!VR)
```
same here, VR and VC created at SetOperators time. Raise error if not present here

stefanozampini commented on linalg/slepc.hpp:
```diff
@@ -0,0 +1,119 @@
+
+   /// Set solver tolerance
+   void SetTol(double tol);
+
```
You may want to add typecasting to get the EPS object (see the PETSc classes) of the SlepcEigensolver

jandrej replied:
Good point.

jetremblay replied:
Isn't what https://github.com/mfem/mfem/blob/45881cbdd5744d3878635803d3a020c8f79223b5/linalg/slepc.hpp#L112 does? Or you also want typecasting to PetscObject?

stefanozampini replied:
Sorry, missed that. Thanks
Since you mentioned, also add the PetscObject conversion to match what is done in petsc.hpp for  the other exposed PETSc classes

jetremblay replied:
I will add that.

mlstowell commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,360 @@
+   bzz->AddDomainIntegrator(new DiffusionIntegrator(u_r_func));
+   bzz->AddDomainIntegrator(new MassIntegrator(e_r_func));
+   bzz->Assemble();
+   btt->EliminateEssentialBCDiag(ess_bdr, numeric_limits<double>::min());
```
I think this is a duplication of line 221

jetremblay replied:
Good catch. Somehow the results were appearing to be correct in my test cases.

mlstowell replied:
The `ess_bdr` array is set to zero so the essential boundary conditions are not being applied in any case.

jetremblay replied:
That is also a typo. Now when run the example correctly with `ess_bdr = 1;`, I get a zero pivot error when running on one process, but it works fine in parallel. Any clue about what is going on?

stefanozampini replied:
No if I do not see a stack trace from the PETSc error. Using a very small number on the diagonal is standard practice in eigenvalue problems?

jetremblay replied:
Using 0 also works in combination with `skip_zeros=0` when calling `Finalize`. I wanted to do something similar to `ex11p`.

jetremblay replied:
This is what I get when I run with `mpirun -np 1`:
```
[0]PETSC ERROR: Zero pivot in LU factorization: https://www.mcs.anl.gov/petsc/documentation/faq.html#zeropivot
[0]PETSC ERROR: Zero pivot row 0 value 2.22507e-305 tolerance 2.22045e-14
```
It does not happen with `mpirun -np 2`

jetremblay replied:
Thanks for the explanation. I will try `skip_zeros = 0` and see if it fixes the problem.

jetremblay replied:
You are right that the serial and parallel LU solvers have different zero pivot checks. I got it to work in serial by using `-st_pc_factor_shift_type NONZERO`. `skip_zeros = 0` did not have any effect in this case.

mlstowell commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,388 @@
+   e_r(0) = -pow(k0*1.0,2);
+   // This is an example to use different refractive indices in mesh domains
+   e_r(1) = -pow(k0*2.0,2);
+   PWConstCoefficient e_r_func(e_r);
```
I believe the default mesh has only one material region so this will always return `e_r(0)`.  It would be good to modify the mesh attributes to create a dielectric slab configuration.  Otherwise this is not really a dielectric waveguide but just a metallic waveguide.

jetremblay replied:
Yes, with only one material this is simulating a metallic waveguide. Should I add an example mesh with two material regions that would resemble a dielectric waveguide? I could only find e.g. `beam-tri.mesh` which is not a very useful example of dielectric waveguide.

mlstowell replied:
It might be easier to reassign selected element attributes like this:
```
   {
      Vector cent(dim);
      for (int i=0; i<pmesh->GetNE(); i++)
      {
         pmesh->GetElementCenter(i, cent);
	 if (fabs(cent[0]-0.5) < 0.125)
	 {
	   pmesh->GetElement(i)->SetAttribute(2);
	 }
      }
   }
```

jetremblay replied:
I added something similar in https://github.com/mfem/mfem/pull/1487/commits/d27f3d683e041b3af4c488f1ad3a818185a8b09a to define a rectangular waveguide core.

mlstowell commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,388 @@
+   }
+
+   // 15. Save data in the VisIt format
+   VisItDataCollection visit_dc("Example5-Parallel", pmesh);
```
Need to update the file name

jetremblay replied:
Done

mlstowell commented on linalg/slepc.cpp:
```diff
@@ -0,0 +1,241 @@
+
+#include "linalg.hpp"
+
+#include "slepc.h"
```
Perhaps this should be `slepc.hpp`.

mlstowell replied:
No, my mistake...

mlstowell commented on examples/petsc/makefile:
```diff
@@ -22,7 +22,7 @@ MFEM_LIB_FILE = mfem_is_not_built
 
 SEQ_EXAMPLES =
-PAR_EXAMPLES = ex1p ex2p ex3p ex4p ex5p ex6p ex9p ex10p
+PAR_EXAMPLES = ex1p ex2p ex3p ex4p ex5p ex6p ex9p ex10p ex11p ex28p
```
These examples should only be added to this list when MFEM_USE_SLEPC is YES.  I imagine something similar may be needed in the CMake build files.

jetremblay replied:
Fixed (both in makefile and CMake build file).

mlstowell commented on config/defaults.mk:
```diff
@@ -276,6 +277,19 @@ ifeq ($(PETSC_FOUND),YES)
 
+SLEPC_ARCH := arch-linux2-c-debug
+SLEPC_DIR := $(MFEM_DIR)/../slepc/$(SLEPC_ARCH)
+SLEPC_VARS := $(SLEPC_DIR)/lib/slepc/conf/slepc_variables
```
I downloaded version 3.13.2 of SLEPC but it doesn't contain this file. It does have `slepcvariables` in this location.  However, this does not seem to work for me.  MFEM's `make config` works without any errors but I cannot build MFEM because my include line doesn't contain SLEPC's include paths.  I've been trying to resolve this but I haven't found the answer yet.  One thing I did find is that `slepcvariables` does not define `SLEPC_INCLUDE`.  Also, `slepcvariables` seems to be a very different type of file from `petscvariables` so perhaps I'm not referencing the correct file here.

mlstowell replied:
Okay, I can build if I replace this path with `$(MFEM_DIR)/../slepc/lib/slepc/conf/slepc_variables` e.g. without the `$(SLEPC_ARCH)`.

jetremblay replied:
Yes this difference between SLEPc and PETSc is strange. I was using my system provided SLEPc installation (which does not define `SLEPC_ARCH`), and I just tried as you did with a vanilla SLEPc installation and `SLEPC_ARCH` is not used either (it reuses `PETSC_ARCH`). I will remove the references to `SLEPC_ARCH`.

jetremblay replied:
@mlstowell please let me know if https://github.com/mfem/mfem/pull/1487/commits/c7fe398bd763a5e5b3458d6e39cd61f781382f21 fixes this.

mlstowell replied:
Yes, I think that fixed it.  Thanks!

mlstowell commented on examples/petsc/ex28p.cpp:
```diff
@@ -0,0 +1,401 @@
+      p_sock.precision(8);
+      p_sock << "solution\n" << *pmesh << *ez << "window_title 'Longitudinal E field'"
+             << endl;
+      p_sock << "keys Rjl!\n";
```
I'm not familiar with the "!" GLVis key but it makes my "Longitudinal E field" plots solid black.  In order to see anything I need to hit 'p' to switch the palette.  In the other plot I don't see an effect.  This may indicate a problem with GLVis on my laptop but you may want to consider removing the '!' unless you really want it.

tzanio replied:
The "!" key switches between discret, smooth and no texture, see https://github.com/GLVis/glvis/blob/master/README#L159-L164.

Its effect is most pronounced in pallets such as 5-color (first choice with `F6`).

If you are getting solid black, that's a bug. Can you report a reproducer on the GLVis issue tracker?

mlstowell replied:
@tzanio did find a bug in GLVis which lead to this problem.  No changes are necessary in this PR.

tzanio commented:
Adding time to finish the review...

jetremblay commented:
I am still having one issue with ex11p when I run in parallel. I seems like part of the solution is corrupted, however the resulting eigenvalue is correct. Am I missing something when extracting the eigenvector in parallel?
![image](https://user-images.githubusercontent.com/15020371/84823488-45aac080-afd3-11ea-8aa9-7a092be580f9.png)

mlstowell commented:
Hi, @jetremblay ,
There's a hidden thing happening when we write `x = lobpcg->GetEigenvector(i);` in `ex11p`. The `HypreLOBPCG` class returns a `HypreParVector` and the `ParGridFunction` recognizes what needs to be done.

Your `SlepcEigenSolver::GetEigenvector` should probably check that the input vector has the correct length.  If you add such a check you should see that the `ParGridFunction` is larger than expected.  To fix `ex11p` you'll need a temporary vector, `temp` for example, of length `fespace->GetTrueVSize()`.  Pass this temporary too SLEPc.  Then distribute those values to the grid function with `x.Distribute(temp);`

Best wishes,
Mark

mlstowell commented:
Hi, @jetremblay ,

There have been a few internal MFEM discussions of this PR in other venues and people are eager to see this become part of MFEM.  One issue that we need to address first is `ex28p`.  The `examples/petsc` subdirectory is intended for PETSc versions of our main examples.  Since we don't have a code like this in `examples` it doesn't really fit in `examples/petsc`.  However, it's a very cool example so I suggested we move it into `miniapps/electromagnetics`.  This would require a few changes but, if you have no objections, I'd be happy to help.

If we do move `ex28p` we'll need to find a new name.  We try to name the miniapps after famous scientists whose work was in some way associated with the problems being solved.  For this one I would suggest `fresnel` for his work on total internal reflection.

Best wishes,
Mark

jetremblay commented:
> Hi, @jetremblay ,
> There's a hidden thing happening when we write `x = lobpcg->GetEigenvector(i);` in `ex11p`. The `HypreLOBPCG` class returns a `HypreParVector` and the `ParGridFunction` recognizes what needs to be done.
> 
> Your `SlepcEigenSolver::GetEigenvector` should probably check that the input vector has the correct length. If you add such a check you should see that the `ParGridFunction` is larger than expected. To fix `ex11p` you'll need a temporary vector, `temp` for example, of length `fespace->GetTrueVSize()`. Pass this temporary too SLEPc. Then distribute those values to the grid function with `x.Distribute(temp);`
> 
> Best wishes,
> Mark

Thanks, this should be fixed now.

jetremblay commented:
> Hi, @jetremblay ,
> 
> There have been a few internal MFEM discussions of this PR in other venues and people are eager to see this become part of MFEM. One issue that we need to address first is `ex28p`. The `examples/petsc` subdirectory is intended for PETSc versions of our main examples. Since we don't have a code like this in `examples` it doesn't really fit in `examples/petsc`. However, it's a very cool example so I suggested we move it into `miniapps/electromagnetics`. This would require a few changes but, if you have no objections, I'd be happy to help.
> 
> If we do move `ex28p` we'll need to find a new name. We try to name the miniapps after famous scientists whose work was in some way associated with the problems being solved. For this one I would suggest `fresnel` for his work on total internal reflection.
> 
> Best wishes,
> Mark

Yes, one of the reasons I wanted to try SLEPc is that the problem in ex28p is not positive definite and LOBPCG was failing. I think a miniapp is a good idea, it could also be made more versatile. Should I remove ex28p from this PR so we can merge it independently?

Regarding the name, I also had in mind `hertz` for his work on electromagnetic waves.

mlstowell commented:
Hi, @jetremblay ,

Great!  We could go either way with the miniapp.  Moving it to a separate PR would push SLEPc into master more quickly.  I'll leave that choice up to you.

Unfortunately, the name `hertz` is already in use for a frequency domain EM code which hasn't made its way into master yet.

Best wishes,
Mark

jetremblay commented:
Hi @mlstowell, I removed ex28p so this PR can move faster.
Regarding the `fresnel` miniapp, I would be glad if you can help. I will have time next week to start working on it.

jandrej commented on CHANGELOG:
```diff
@@ -96,6 +98,13 @@ New and updated examples and miniapps
+- Ported example 11p to SLEPc, to demonstrate solving the Laplace eigenvalue
+  equation with the shift-and-invert spectral transformation method.
+
+- Added a new example, Example 28p, to demonstrate the solution of the 2D
+  Maxwell eigenvalue problem in a dielectric waveguide. This examples shows the
+  use of the SLEPc eigensolvers with mixed finite elements and block operators.
+
```
Still mentions the soon to be miniapp (ex28p)

jetremblay replied:
Fixed

jandrej commented on examples/petsc/ex11p.cpp:
```diff
@@ -0,0 +1,461 @@
+//
+// Compile with: make ex11p
+//
+// Sample runs:  mpirun -np 4 ex11p -m ../../data/square-disc.mesh
+//               mpirun -np 4 ex11p -m ../../data/star.mesh
+//               mpirun -np 4 ex11p -m ../../data/star-mixed.mesh
```
Could you update the sample runs to only petsc options? This is something for our test system and those are already tested in the regular run. Just have a few different options which only use petsc and one that doesn't.

jetremblay replied:
I updated the sample runs, and added a `rc_ex11p` file to test the `--slepcopts` parameter.

jandrej commented:
I ran tested examples and left a few minor comments. Imho this is ready.

mlstowell commented:
@jetremblay , I look forward to your next PR and I'l be glad to help!

tzanio commented:
@stefanozampini do you have any outstanding concerns with this PR?

stefanozampini commented on config/cmake/modules/FindSLEPc.cmake:
```diff
@@ -0,0 +1,278 @@
+  ierr = SlepcInitialize(&argc, &argv, PETSC_NULL, PETSC_NULL);
+  EPS eps;
+  ierr = EPSCreate(PETSC_COMM_SELF, &eps); CHKERRQ(ierr);
+  //ierr = EPSSetFromOptions(eps); CHKERRQ(ierr);
```
why this commented line?

jetremblay replied:
Not sure, I will try to simplify this file.

stefanozampini commented:
@tzanio @jetremblay I only have one comment: can we squash commits to have a meaningful history and possibly commits that compiles fine?

stefanozampini commented:
Another comment: from my slepc_variables
```
SLEPC_EXTERNAL_LIB = ${ARPACK_LIB} ${BLOPEX_LIB} ${BLZPACK_LIB} ${HPDDM_LIB} ${PRIMME_LIB} ${SLICOT_LIB} ${TRLAN_LIB}
```

This is not telling us anything about where to look for the slepc lib (In PETSc instead, PETSC_EXTERNAL_LIB has hardcoded paths)

This is what I have in my config/user.mk to have it pick the proper link
```
...
SLEPC_LIB = dummy
include path_to_slepcvariables
```

stefanozampini commented:
I'm getting compilation warnings with -Wmaybe-uninitialized. Also, can you add the rule `ex11p-test-par` using the rc_file? it seems the rc_file is not tested

```
diff --git a/examples/petsc/ex11p.cpp b/examples/petsc/ex11p.cpp
index a298ab864..c3be0128a 100644
--- a/examples/petsc/ex11p.cpp
+++ b/examples/petsc/ex11p.cpp
@@ -278,8 +278,8 @@ int main(int argc, char *argv[])
       }
    }
 
-   HypreLOBPCG * lobpcg;
-   SlepcEigenSolver * slepc;
+   HypreLOBPCG * lobpcg = NULL;
+   SlepcEigenSolver * slepc = NULL;
    if (!use_slepc)
    {
```

stefanozampini commented:
Just opened a merge request in PETSc to automatically configure MFEM with SLEPc support if PETSc is configured with `--download-mfem --download-mfem-git-commit=slepc-dev --download-slepc` https://gitlab.com/petsc/petsc/-/merge_requests/2892

stefanozampini commented:
You may want to include an example on how to use LOBPCG from SLEPc also
 
```
-eps_monitor
-eps_converged_reason
-st_type precond
-eps_view_values
-eps_smallest_real
-eps_type lobpcg
-eps_gen_hermitian
-eps_lobpcg_blocksize 5
-st_pc_type gamg
```

this produces the nice output below

```
[szampini@localhost petsc]$ ./ex11p --slepcopts rc_ex11p -no-vis 
Options used:
   --mesh ../../data/star.mesh
   --refine-serial 2
   --refine-parallel 1
   --order 1
   --num-eigs 5
   --seed 75
   --no-visualization
   --useslepc
   --slepcopts rc_ex11p
Number of unknowns: 1361
  Eigenvalue approximations and residual norms for  solve.
  1 EPS nconv=0 first unconverged value (error) 4.74693 (4.76208832e-02)
  2 EPS nconv=0 first unconverged value (error) 4.5415 (4.03174229e-03)
  3 EPS nconv=0 first unconverged value (error) 4.54055 (1.88104407e-04)
  4 EPS nconv=0 first unconverged value (error) 4.54055 (6.10515009e-06)
  5 EPS nconv=0 first unconverged value (error) 4.54055 (1.47412677e-07)
  6 EPS nconv=1 first unconverged value (error) 11.3202 (1.59019260e-05)
  7 EPS nconv=1 first unconverged value (error) 11.3202 (2.81783560e-06)
  8 EPS nconv=1 first unconverged value (error) 11.3202 (2.83426849e-07)
  9 EPS nconv=1 first unconverged value (error) 11.3202 (3.13103473e-08)
 10 EPS nconv=3 first unconverged value (error) 19.2291 (3.67497437e-06)
 11 EPS nconv=3 first unconverged value (error) 19.2291 (6.11334764e-07)
 12 EPS nconv=3 first unconverged value (error) 19.2291 (2.03211216e-07)
 13 EPS nconv=3 first unconverged value (error) 19.2291 (8.53101663e-08)
 14 EPS nconv=3 first unconverged value (error) 19.2291 (3.53203068e-08)
 15 EPS nconv=3 first unconverged value (error) 19.2291 (1.16765598e-08)
 Linear eigensolve converged (5 eigenpairs) due to CONVERGED_TOL; iterations 16
Eigenvalues = 
   4.54055
   11.32017
   11.32018
   19.22912
   19.22915


```

tzanio commented:
> @tzanio @jetremblay I only have one comment: can we squash commits to have a meaningful history and possibly commits that compiles fine?

Thanks @stefanozampini. I'll squash and merge once @jeremylt address the above comments.

tzanio commented:
LGTM

tzanio commented:
Merged in `next` for testing...

jetremblay commented:
> Merged in `next` for testing...

@tzanio can I still push to this branch to address the last comments, even if it was merged in `next`?

tzanio commented:
Yes, please address the issues and push and I will re-merge in `next`.

jetremblay commented:
Hopefully https://github.com/mfem/mfem/pull/1487/commits/68ecd01b9a72c6ef8c68838554eb914ec626163f addresses all the remaining issues.

- The FindSLEPc.cmake file was simplified based on `mfem_find_package`. It works on my system but I am not sure how robust it is.
- The makefile now sources `slepcvariables` (in addition to `slepc_variables`) to know about the SLEPc external libraries. I tested this with SLEPc built with ARPACK.
- Indeed I forgot the test targets for `petsc/ex11p`. This should now work both with make and cmake.
- I added `rc_ex11p_lobpcg` as suggested to test LOBPCG through SLEPc.

Thanks everyone for your help!

tzanio commented:
Re-merged in `next` for testing ...

tzanio commented:
LGTM

tzanio commented:
LGTM
